plugin/pkg/admission/limitranger/admission.go

```diff
@@ -161,21 +163,23 @@ func (l *LimitRanger) GetLimitRanges(a admission.Attributes) ([]*corev1.LimitRan
 	if len(items) == 0 {
 		lruItemObj, ok := l.liveLookupCache.Get(a.GetNamespace())
 		if !ok || lruItemObj.(liveLookupEntry).expiry.Before(time.Now()) {
-			// TODO: If there are multiple operations at the same time and cache has just expired,
-			// this may cause multiple List operations being issued at the same time.
-			// If there is already in-flight List() for a given namespace, we should wait until
-			// it is finished and cache is updated instead of doing the same, also to avoid
-			// throttling - see #22422 for details.
-			liveList, err := l.client.CoreV1().LimitRanges(a.GetNamespace()).List(context.TODO(), metav1.ListOptions{})
+			// Fixed: #22422
+			// use singleflight to alleviate simultaneous calls to
+			lruItemObj, err, _ = l.group.Do(a.GetNamespace(), func() (interface{}, error) {
```

Comment:
if I switch this to partition on a constant string "x" instead of the namespace, the added unit test still passes, so the test is not effectively verifying this is correct

Reply:
If the constant string x will be used, then he will partition all calls. If there are two concurrent calls with different namespaces, he will only call them once until the call returns. This should be the wrong behavior.

Currently using namespaces as partitions, this is partitioning a set of calls with the same namespace. This is the same behavior as using namespaces as keys for LRU caching before
@liggitt

Reply:
using a constant string "x" here would be a bug, since it would make calls from different namespaces share a single lookup and use the same retrieved result. The unit test should fail under those circumstances, and it does not.

Reply:
If you use the constant string x, the test will block

Reply:
that's not what I'm seeing:

```shell
git diff
```

```diff
diff --git a/plugin/pkg/admission/limitranger/admission.go b/plugin/pkg/admission/limitranger/admission.go
index 4525463d2e6..907a28bda32 100644
--- a/plugin/pkg/admission/limitranger/admission.go
+++ b/plugin/pkg/admission/limitranger/admission.go
@@ -165,7 +165,7 @@ func (l *LimitRanger) GetLimitRanges(a admission.Attributes) ([]*corev1.LimitRan
                if !ok || lruItemObj.(liveLookupEntry).expiry.Before(time.Now()) {
                        // Fixed: #22422
                        // use singleflight to alleviate simultaneous calls to
-                       lruItemObj, err, _ = l.group.Do(a.GetNamespace(), func() (interface{}, error) {
+                       lruItemObj, err, _ = l.group.Do("x", func() (interface{}, error) {
                                liveList, err := l.client.CoreV1().LimitRanges(a.GetNamespace()).List(context.TODO(), metav1.ListOptions{})
                                if err != nil {
                                        return nil, admission.NewForbidden(a, err)
```

```shell
go test ./plugin/pkg/admission/limitranger -count=1 -v -run TestLimitRanger_GetLimitRangesFixed22422
=== RUN   TestLimitRanger_GetLimitRangesFixed22422
--- PASS: TestLimitRanger_GetLimitRangesFixed22422 (0.00s)
PASS
ok  	k8s.io/kubernetes/plugin/pkg/admission/limitranger	0.800s
```

Reply:
I just updated it and the problem should be fixed

Reply:
go test ./plugin/pkg/admission/limitranger -count=1 -v -run TestLimitRanger_GetLimitRangesFixed22422
=== RUN   TestLimitRanger_GetLimitRangesFixed22422
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
    admission_test.go:910: Expected test namespace, got test1
--- FAIL: TestLimitRanger_GetLimitRangesFixed22422 (0.00s)
FAIL
FAIL    k8s.io/kubernetes/plugin/pkg/admission/limitranger      0.792s
FAIL